fix for ICE issue #25180

llogiq opened the pull request:
According to @eddyb – and my tests – the following gets rid of the ICE in issue #25180.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @pnkfelix (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. The way Github handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [CONTRIBUTING.md](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

llogiq commented:
I think what the PR lacks as of yet is a testcase to show that the fix actually quells the ICE. I just added it and will commit after I rebuild.

eddyb commented on src/librustc/middle/check_const.rs:
```diff
@@ -251,9 +251,13 @@ impl<'a, 'tcx, 'v> Visitor<'v> for CheckCrateVisitor<'a, 'tcx> {
-        assert!(self.mode == Mode::Var);
+        let (old_mode, old_qualif) = (self.mode, self.qualif);
+        self.mode = Mode::Var;
+        self.qualif = ConstQualif::empty();
```
You don't have to *copy* `with_mode`, you can just call it (and replace `self` in the two original statements with `v`, or whatever you choose for the name of the closure argument).

llogiq commented:
Yep, you're right. I'll change it.

eddyb commented on src/librustc/middle/check_const.rs:
```diff
@@ -251,8 +251,8 @@ impl<'a, 'tcx, 'v> Visitor<'v> for CheckCrateVisitor<'a, 'tcx> {
-        self.with_euv(Some(fn_id), |euv| euv.walk_fn(fd, b));
+        self.with_mode(Mode::Var, |v| v.with_euv(Some(fn_id),
+            |euv| euv.walk_fn(fd, b)));
         visit::walk_fn(self, fk, fd, b, s);
```
`walk_fn` has to be inside the `with_mode` call, too.

llogiq commented:
OK, but I'd like to understand why? Surely rustc compiles the code anyway – I even have a test case saying so. What are the implications of walking the fn in const mode?

eddyb commented:
If it's in const mode, you get the const rules. And maybe ICEs. Try calling a function and you'll see.

llogiq commented:
Ah, Ok. Should have run the extended test before I fixed the function.

Is this ready to merge now or do I need to do something else?

eddyb commented on src/test/run-pass/issue-25180.rs:
```diff
@@ -0,0 +1,18 @@
+
+// pretty-expanded FIXME #25180
+
+const x: &'static Fn() = &|| println!("ICE here");
```
Do multiple statements, arguments and local variables work in these closures?

llogiq commented:
Yeah, it's a good idea to add some more tests. I'm on it.

llogiq commented:
So yes, it [appears to work](https://github.com/llogiq/rust/blob/master/src/test/run-pass/issue-25180.rs).

llogiq commented:
I guess I'll try and fix up the git history before you pull.

eddyb commented:
r? @eddyb @bors r+ rollup

bors commented:
:pushpin: Commit a18ce4d has been approved by `eddyb`

<!-- @bors r=eddyb a18ce4d47979260001e9be0a92f887e3abd4eb88 -->

bors commented:
:hourglass: Testing commit a18ce4d with merge b758a37...

bors commented:
:broken_heart: Test failed - [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/4891)

eddyb commented on src/test/run-pass/issue-25180.rs:
```diff
@@ -0,0 +1,30 @@
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// pretty-expanded FIXME #25180
```
The failure was in pretty tests... Which I don't you can do if you're using macros like `println!` which use unstable items in their expansions..

steveklabnik commented:
@bors: rollup-

eddyb commented:
@steveklabnik huh, shouldn't a test failure take it out of the queue automatically?

steveklabnik commented:
don't think so

llogiq commented:
I'm sorry that I broke the rollup. Apparently my tests weren't sufficient. I've now removed the print! macro invocations and added a feature declaration (which was hinted in the error message).

Can someone tell me how I can reproduce the failing test?

steveklabnik commented:
It's all good :) everyone does it. I'm not 100% sure how to reproduce though

llogiq commented:
Thanks for the kind words. My first PR and manage to hit all available roadblocks at once.

But hey, I learn a lot there days. :smile:

eddyb commented on src/test/run-pass/issue-25180.rs:
```diff
@@ -0,0 +1,30 @@
+
+const EMPTY: &'static Fn() = &|| ();
+
+const ONE_ARGUMENT: &'static Fn(u32) = &|y| assert!(y == 42);
```
`assert!` still uses the formatting machinery. I'm curious if we prefer to disable pretty testing or just add feature gates. What do other tests do?

llogiq commented:
I see 783 instances of "feature" just in `test/run-pass`. So yes, it appears feature-gates are not that uncommon. I'm unsure if #![feature(std_misc)] will be enough to make it work.

eddyb commented:
@llogiq Running `make check-stage1-rpass-pretty` should answer that question.

llogiq commented:
@eddyb Thanks, but that only prints that my Makefile does not have this target. Perhaps I'm missing some required package?

pnkfelix commented:
@llogiq no I think @eddyb just mispoke.  The target is (probably) `make check-stage1-pretty-rpass`.

(I always recommend GNU remake for debugging big makefile systems.)

eddyb commented:
@llogiq @pnkfelix My bad, I have both in my shell history and I pasted the bad one.

llogiq commented:
In that case, my previous two commits make the test pass on my machine.

bors commented:
:umbrella: The latest upstream changes (presumably #25609) made this pull request unmergeable. Please resolve the merge conflicts.

llogiq commented:
I've just revisited the problem, and recent changes not only made my PR unmergeable, but also fixed the problem in a different way. Closing the PR is the sensible option here.
